catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswSubscriptionEndpoint.java

```diff
@@ -0,0 +1,546 @@
+        // to registry
+        if (registeredSubscriptions.containsKey(subscriptionUuid)) {
+            LOGGER.debug("Delete existing subscription {} for re-creation", subscriptionUuid);
+            deleteSubscription(subscriptionUuid);
+        }
+        CswSubscription sub = createSubscription(request);
+        LOGGER.debug("Registering Subscription");
+        ServiceRegistration serviceRegistration =
+                context.registerService(Subscription.class.getName(), sub, props);
+
+        if (serviceRegistration != null) {
+            LOGGER.debug("Subscription registered with bundle ID = {} ",
+                    serviceRegistration.getReference()
+                            .getBundle()
+                            .getBundleId());
+            registeredSubscriptions.put(subscriptionUuid, serviceRegistration);
+            // Pass in client-provided subscriptionId vs. subscription UUID because
+            // the filter XML to be persisted for this subscription will be used to
+            // restore this subscription and should consist of the exact values the
+            // client originally provided.
+            if (persistSubscription) {
+                persistSubscription(sub, deliveryMethodUrl, subscriptionUuid);
+            }
+        } else {
+            LOGGER.debug("Subscription registration failed");
+        }
+
+        LOGGER.debug("EXITING: {}", methodName);
+        return subscriptionUuid;
+
+    }
+
+    private synchronized boolean deleteSubscription(String subscriptionId) throws CswException {
+        String methodName = "deleteSubscription";
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("ENTERING: {}", methodName);
+            LOGGER.debug("subscriptionId = {}", subscriptionId);
+        }
+
+        if (StringUtils.isEmpty(subscriptionId)) {
+            throw new CswException(
+                    "Unable to delete subscription because subscription ID is null or empty");
+        }
+
+        boolean status = false;
+
+        try {
+            LOGGER.debug("Removing (unregistering) subscription: {}", subscriptionId);
+            ServiceRegistration sr = (ServiceRegistration) registeredSubscriptions.remove(
+                    subscriptionId);
+            if (sr != null) {
+                sr.unregister();
+            } else {
+                LOGGER.debug("No ServiceRegistration found for subscription: {}", subscriptionId);
+            }
+
+            Configuration subscriptionConfig = getSubscriptionConfiguration(subscriptionId);
+            try {
+                if (subscriptionConfig != null) {
+                    LOGGER.debug("Deleting subscription for subscriptionId = {}", subscriptionId);
+                    subscriptionConfig.delete();
+
+                    // Subscription removal is only successful if able to remove from OSGi registry and
+                    // ConfigAdmin service
+                    if (sr != null) {
```

Comment:
if we can't delete from ConfigAdmin and they get out of sync then things are terribly wrong and will stay out of sync until that is fixed and the bundle is restarted. Once the bundle has been restarted the managed service factory will recreate all of the services from config and populate the hashmap.  Yes, I agree the config stuff really could be abstracted out but its a low priority until we address some of the other existing issues with how events are handled.